core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java

```diff
@@ -0,0 +1,777 @@
+                            throw new MapperParsingException("error parsing field [" +
+                                name() + "], with unknown parameter [" + fieldName + "]");
+                        }
+                    }
+                }
+                range = new Range(from, to, includeFrom, includeTo);
+            } else {
+                throw new MapperParsingException("error parsing field ["
+                    + name() + "], expected an object but got " + parser.currentName());
+            }
+        }
+        if (includeInAll) {
+            context.allEntries().addText(fieldType.name(), range.toString(), fieldType.boost());
+        }
+        boolean indexed = fieldType.indexOptions() != IndexOptions.NONE;
+        boolean docValued = fieldType.hasDocValues();
+        boolean stored = fieldType.stored();
+        fields.addAll(fieldType().rangeType.createFields(name(), range, indexed, docValued, stored));
+    }
+
+    @Override
+    protected void doMerge(Mapper mergeWith, boolean updateAllTypes) {
+        super.doMerge(mergeWith, updateAllTypes);
+        RangeFieldMapper other = (RangeFieldMapper) mergeWith;
+        this.includeInAll = other.includeInAll;
+        if (other.coerce.explicit()) {
+            this.coerce = other.coerce;
+        }
+    }
+
+    @Override
+    protected void doXContentBody(XContentBuilder builder, boolean includeDefaults, Params params) throws IOException {
+        super.doXContentBody(builder, includeDefaults, params);
+
+        if (includeDefaults || (fieldType().dateTimeFormatter() != null
+                && fieldType().dateTimeFormatter().format().equals(DateFieldMapper.DEFAULT_DATE_TIME_FORMATTER.format()) == false)) {
+            builder.field("format", fieldType().dateTimeFormatter().format());
+        }
+        if (includeDefaults || (fieldType().dateTimeFormatter() != null
+                && fieldType().dateTimeFormatter().locale() != Locale.ROOT)) {
+            builder.field("locale", fieldType().dateTimeFormatter().locale());
+        }
+        if (includeDefaults || coerce.explicit()) {
+            builder.field("coerce", coerce.value());
+        }
+        if (includeInAll != null) {
+            builder.field("include_in_all", includeInAll);
+        } else if (includeDefaults) {
+            builder.field("include_in_all", false);
+        }
+    }
+
+    /** Enum defining the type of range */
+    public enum RangeType {
+        DATE("date_range", NumberType.LONG) {
+            @Override
+            public Field getRangeField(String name, Range r) {
+                return new LongRangeField(name, new long[] {(Long)r.from}, new long[] {(Long)r.to});
+            }
+            @Override
+            public Number parse(RangeFieldType fieldType, XContentParser parser, boolean coerce, LongSupplier nowInMillis)
+                    throws IOException {
+                return fieldType.dateMathParser.parse(parser.text(), nowInMillis);
```

Comment:
should we round the lower bound down and the upper bound up like DateFieldType.innerRangeQuery?
Also I think you could use a `nowInMillis` impl that just throws an exception since `now` is not supposed to be used at index time.